### PR TITLE
Add aten::flatten NNAPI converter

### PR DESCRIPTION
**Pull request by axitkhurana**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61025
* #61024
* #61023
* #61022
* #61021
* #59529
* #59434
* #59365
* #59364
* #58543
* **#60885**
* #58541
* #58540
* #58539

Summary:

Add support for aten::flatten op in the NNAPI model converter.
Variable size support to supported soon

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_flatten

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29451725](https://our.internmc.facebook.com/intern/diff/D29451725)

### CODE REVIEW

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

File: torch/backends/_nnapi/serializer.py [hunk 1]
```diff
@@ -918,6 +921,53 @@ def add_reshape(self, node):
...
         self.add_operation(NNAPI_OperationCode.RESHAPE, inputs, outputs)
 
+    def add_flatten(self, node):
+        assert node.inputsSize() == 3
+        assert node.outputsSize() == 1
+
+        in_id, in_oper = self.get_tensor_operand_by_jitval_fixed_size(node.inputsAt(0))
+
+        start_ctype, start_dim = self.get_constant_value(node.inputsAt(1))
+        end_ctype, end_dim = self.get_constant_value(node.inputsAt(2))
```
Comment (dreiss) [inline torch/backends/_nnapi/serializer.py:931]: Force these to be ints, like you did for softmax dim?

File: torch/backends/_nnapi/serializer.py [hunk 2]
```diff
@@ -918,6 +921,53 @@ def add_reshape(self, node):
...
+        #     raise Exception("Flattened dims can't be flexible")
+        # non_flattened_dims = in_oper.shape[: start_dim] + in_oper.shape[end_dim + 1:]
+        # if non_flattened_dims.count(0) > 1:
+        #     raise Exception("Only 1 dim can be flexible")
+        # out_shape = tuple(
+        #     dim if dim != 0 else -1
+        #     for dim in out_shape
+        # )
+
+        out_oper = in_oper._replace(shape=out_shape, dim_order=DimOrder.PRESUMED_CONTIGUOUS)
```
Comment (dreiss) [inline torch/backends/_nnapi/serializer.py:959]: Order is already contiguous.

**Review by dreiss — APPROVED** [review]

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

**Pull request closed without merging**